discoverd/server/store_test.go

```diff
@@ -0,0 +1,619 @@
+		t.Fatal(err)
+	} else if err = s.SetServiceMeta("service0", &discoverd.ServiceMeta{Data: []byte(`"foo"`), Index: 0}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Update metadata with wrong previous index.
+	if err := s.SetServiceMeta("service0", &discoverd.ServiceMeta{Data: []byte(`"bar"`), Index: 100}); err == nil || err.Error() != `precondition_failed: Service metadata for "service0" exists, but wrong index provided` {
+		t.Fatalf("unexpected error: %v", err)
+	}
+}
+
+// Ensure the store returns an error when setting metadata for a non-existent service.
+func TestStore_SetServiceMeta_ErrNotFound(t *testing.T) {
+	s := MustOpenStore()
+	defer s.Close()
+	if err := s.SetServiceMeta("service0", &discoverd.ServiceMeta{Data: []byte(`"foo"`), Index: 0}); !server.IsNotFound(err) {
+		t.Fatalf("unexpected error: %s", err)
+	}
+}
+
+// Ensure the store can manually set a leader for a manual service.
+func TestStore_SetLeader(t *testing.T) {
+	s := MustOpenStore()
+	defer s.Close()
+	if err := s.AddService("service0", &discoverd.ServiceConfig{LeaderType: discoverd.LeaderTypeManual}); err != nil {
+		t.Fatal(err)
+	} else if err := s.AddInstance("service0", &discoverd.Instance{ID: "inst0"}); err != nil {
+		t.Fatal(err)
+	} else if err := s.AddInstance("service0", &discoverd.Instance{ID: "inst1"}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Set the leader instance ID.
+	if err := s.SetLeader("service0", "inst1"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Verify that the leader was set.
+	if inst := s.Leader("service0"); !reflect.DeepEqual(inst, &discoverd.Instance{ID: "inst1"}) {
+		t.Fatalf("unexpected leader: %#v", inst)
+	}
+}
+
+// Ensure the store does not error when setting a leader for a non-existent service.
+func TestStore_SetLeader_NoService(t *testing.T) {
+	s := MustOpenStore()
+	defer s.Close()
+	if err := s.SetLeader("service0", "inst1"); err != nil {
+		t.Fatal(err)
+	} else if inst := s.Leader("service0"); inst != nil {
+		t.Fatalf("unexpected leader: %#v", inst)
+	}
+}
+
+// Ensure the store does not error when setting a leader for a non-existent instance.
+func TestStore_SetLeader_NoInstance(t *testing.T) {
+	s := MustOpenStore()
+	defer s.Close()
+	if err := s.AddService("service0", &discoverd.ServiceConfig{LeaderType: discoverd.LeaderTypeManual}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := s.SetLeader("service0", "inst1"); err != nil {
+		t.Fatal(err)
+	} else if inst := s.Leader("service0"); inst != nil {
+		t.Fatalf("unexpected leader: %#v", inst)
+	}
+}
+
+// Store represents a test wrapper for server.Store.
+type Store struct {
+	*server.Store
+}
+
+// NewStore returns a new instance of Store.
+func NewStore() *Store {
+	// Generate a temporary path.
+	f, _ := ioutil.TempFile("", "discoverd-store-")
+	f.Close()
+	os.Remove(f.Name())
+
+	// Initialize store.
+	s := &Store{Store: server.NewStore(f.Name())}
+
+	// Set default test settings.
+	s.BindAddress = ":20000"
+	s.Advertise, _ = net.ResolveTCPAddr("tcp", "localhost:20000")
```

Comment:
Does that safely work for you? I've had issues where it'll reassign the port after the `net.Listener` is closed. Although that may have only been an issue because I was running `go test` with `-parallel`.